<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" >

    <!-- Include other parts of the robot here -->

    <!-- INFRARED FRAME LINK -->
    <link name="infrared_frame">
        <visual>
            <geometry>
                <box size="0.15 0.01 0.01"/> <!-- A bar to hold the infrared sensors -->
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.15 0.01 0.01"/>
            </geometry>
        </collision>
        <xacro:inertial_box mass="0.05" x="0.15" y="0.01" z="0.01">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_box>
    </link>

    <joint name="infrared_joint" type="fixed">
        <parent link="chassis"/>
        <child link="infrared_frame"/>
        <origin xyz="0.29 0 0" rpy="0 0 -${pi/2}" />
    </joint>

    <!-- INFRARED SENSORS -->

    <!-- Infrared Sensor 1 -->
    <link name="infrared_sensor_1">
        <visual>
            <geometry>
                <cylinder radius="0.005" length="0.02"/>
            </geometry>
            <material name="red"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.005" length="0.02"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.01" length="0.02" radius="0.005">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <joint name="infrared_sensor_1_joint" type="fixed">
        <parent link="infrared_frame"/>
        <child link="infrared_sensor_1"/>
        <origin xyz="-0.06 0 0" rpy="0 0 0"/>
    </joint>

    <!-- Infrared Sensor 2 -->
    <link name="infrared_sensor_2">
        <visual>
            <geometry>
                <cylinder radius="0.005" length="0.02"/>
            </geometry>
            <material name="red"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.005" length="0.02"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.01" length="0.02" radius="0.005">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <joint name="infrared_sensor_2_joint" type="fixed">
        <parent link="infrared_frame"/>
        <child link="infrared_sensor_2"/>
        <origin xyz="-0.03 0 0" rpy="0 0 0"/>
    </joint>

    <!-- Infrared Sensor 3 (Center) -->
    <link name="infrared_sensor_3">
        <visual>
            <geometry>
                <cylinder radius="0.005" length="0.02"/>
            </geometry>
            <material name="red"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.005" length="0.02"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.01" length="0.02" radius="0.005">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <joint name="infrared_sensor_3_joint" type="fixed">
        <parent link="infrared_frame"/>
        <child link="infrared_sensor_3"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>

    <!-- Infrared Sensor 4 -->
    <link name="infrared_sensor_4">
        <visual>
            <geometry>
                <cylinder radius="0.005" length="0.02"/>
            </geometry>
            <material name="red"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.005" length="0.02"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.01" length="0.02" radius="0.005">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <joint name="infrared_sensor_4_joint" type="fixed">
        <parent link="infrared_frame"/>
        <child link="infrared_sensor_4"/>
        <origin xyz="0.03 0 0" rpy="0 0 0"/>
    </joint>

    <!-- Infrared Sensor 5 -->
    <link name="infrared_sensor_5">
        <visual>
            <geometry>
                <cylinder radius="0.005" length="0.02"/>
            </geometry>
            <material name="red"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.005" length="0.02"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.01" length="0.02" radius="0.005">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <joint name="infrared_sensor_5_joint" type="fixed">
        <parent link="infrared_frame"/>
        <child link="infrared_sensor_5"/>
        <origin xyz="0.06 0 0" rpy="0 0 0"/>
    </joint>

    <!-- INFRARED SENSOR 1 PLUGIN -->
    <gazebo reference="infrared_sensor_1">
        <material>Gazebo/Red</material>

        <sensor name="infrared_1" type="ray">
            <pose> 0 0 0 0 0 0 </pose>
            <visualize>true</visualize>
            <update_rate>20</update_rate>
            <ray>
                <scan>
                    <horizontal>
                        <samples>1</samples>
                        <min_angle>0.0</min_angle>
                        <max_angle>0.0</max_angle>
                    </horizontal>
                </scan>
                <range>
                    <min>0.01</min>
                    <max>0.1</max> <!-- Short-range suitable for line-following -->
                </range>
            </ray>
            <plugin name="infrared_1_controller" filename="libgazebo_ros_ray_sensor.so">
                <ros>
                    <argument>~/out:=infrared_1</argument>
                </ros>
                <output_type>sensor_msgs/LaserScan</output_type>
                <frame_name>infrared_sensor_1</frame_name>
            </plugin>
        </sensor>
    </gazebo>

    <!-- INFRARED SENSOR 2 PLUGIN -->
    <gazebo reference="infrared_sensor_2">
        <material>Gazebo/Red</material>

        <sensor name="infrared_2" type="ray">
            <pose> 0 0 0 0 0 0 </pose>
            <visualize>true</visualize>
            <update_rate>20</update_rate>
            <ray>
                <scan>
                    <horizontal>
                        <samples>1</samples>
                        <min_angle>0.0</min_angle>
                        <max_angle>0.0</max_angle>
                    </horizontal>
                </scan>
                <range>
                    <min>0.01</min>
                    <max>0.1</max>
                </range>
            </ray>
            <plugin name="infrared_2_controller" filename="libgazebo_ros_ray_sensor.so">
                <ros>
                    <argument>~/out:=infrared_2</argument>
                </ros>
                <output_type>sensor_msgs/LaserScan</output_type>
                <frame_name>infrared_sensor_2</frame_name>
            </plugin>
        </sensor>
    </gazebo>

    <!-- INFRARED SENSOR 3 PLUGIN -->
    <gazebo reference="infrared_sensor_3">
        <material>Gazebo/Red</material>

        <sensor name="infrared_3" type="ray">
            <pose> 0 0 0 0 0 0 </pose>
            <visualize>true</visualize>
            <update_rate>20</update_rate>
            <ray>
                <scan>
                    <horizontal>
                        <samples>1</samples>
                        <min_angle>0.0</min_angle>
                        <max_angle>0.0</max_angle>
                    </horizontal>
                </scan>
                <range>
                    <min>0.01</min>
                    <max>0.1</max>
                </range>
            </ray>
            <plugin name="infrared_3_controller" filename="libgazebo_ros_ray_sensor.so">
                <ros>
                    <argument>~/out:=infrared_3</argument>
                </ros>
                <output_type>sensor_msgs/LaserScan</output_type>
                <frame_name>infrared_sensor_3</frame_name>
            </plugin>
        </sensor>
    </gazebo>

    <!-- INFRARED SENSOR 4 PLUGIN -->
    <gazebo reference="infrared_sensor_4">
        <material>Gazebo/Red</material>

        <sensor name="infrared_4" type="ray">
            <pose> 0 0 0 0 0 0 </pose>
            <visualize>true</visualize>
            <update_rate>20</update_rate>
            <ray>
                <scan>
                    <horizontal>
                        <samples>1</samples>
                        <min_angle>0.0</min_angle>
                        <max_angle>0.0</max_angle>
                    </horizontal>
                </scan>
                <range>
                    <min>0.01</min>
                    <max>0.1</max>
                </range>
            </ray>
            <plugin name="infrared_4_controller" filename="libgazebo_ros_ray_sensor.so">
                <ros>
                    <argument>~/out:=infrared_4</argument>
                </ros>
                <output_type>sensor_msgs/LaserScan</output_type>
                <frame_name>infrared_sensor_4</frame_name>
            </plugin>
        </sensor>
    </gazebo>

    <!-- INFRARED SENSOR 5 PLUGIN -->
    <gazebo reference="infrared_sensor_5">
        <material>Gazebo/Red</material>

        <sensor name="infrared_5" type="ray">
            <pose> 0 0 0 0 0 0 </pose>
            <visualize>true</visualize>
            <update_rate>20</update_rate>
            <ray>
                <scan>
                    <horizontal>
                        <samples>1</samples>
                        <min_angle>0.0</min_angle>
                        <max_angle>0.0</max_angle>
                    </horizontal>
                </scan>
                <range>
                    <min>0.01</min>
                    <max>0.1</max>
                </range>
            </ray>
            <plugin name="infrared_5_controller" filename="libgazebo_ros_ray_sensor.so">
                <ros>
                    <argument>~/out:=infrared_5</argument>
                </ros>
                <output_type>sensor_msgs/LaserScan</output_type>
                <frame_name>infrared_sensor_5</frame_name>
            </plugin>
        </sensor>
    </gazebo>

</robot>
